### ISSUE TITLE
states.file.managed and modules.file.manage_file fail on some http downloads

### ISSUE THREAD

**Issue by jdelic** [saltstack/salt#44373]
### Description of Issue/Question
Downloading a file using `file.managed` from a local http server (Apache2) which previously (in versions <2017.7.2, probably 2016.11.*) succeeded, now fails. It results in a 0 bytes file which subsequently fails the hash check, failing the state. Interestingly, downloading the same file from github.com works. It seems to be a bug in `salt.utils http.query` (tornado backend) or `salt.modules.file.manage_file` since 2016.11.

### Setup
Apache2 fileserver config:
```
<VirtualHost *:80>
    ServerName fileserver.maurusnet.test
    DocumentRoot /srv/aptly/public
    Alias /downloads /srv/downloads
    IndexOptions FancyIndexing
    <Directory /srv/downloads/>
        Require all granted
        Options Indexes
    </Directory>
</VirtualHost>
```

`/srv/downloads/concourse/concourse_linux_amd64` contains a copy of [concourse CI v3.6.0](http://concourse.ci/downloads.html#v360) [concourse_linux_amd64](https://github.com/concourse/concourse/releases/download/v3.6.0/concourse_linux_amd64).

This is downloaded in a salt state like this:
```
concourse-user:
    group.present:
        - name: concourse
    user.present:
        - name: concourse
        - gid: concourse
        - groups:
            - gpg-access
        - optional_groups:
            - docker
        - home: /etc/concourse
        - shell: /bin/false
        - createhome: False
        - require:
            - group: concourse-user

concourse-install:
    file.managed:
        - name: /usr/local/bin/concourse_linux_amd64
        - source: http://fileserver.maurusnet.test/downloads/concourse/concourse_linux_amd64
        - source_hash: sha256=87132e62f227e776b10178c21c83eb9f0049e4b53adc4e4dbb7eb06dfbb2e3b1
        - mode: '0755'
        - user: concourse
        - group: concourse
        - require:
            - user: concourse-user
```

This state fails like this:
```
----------
          ID: concourse-install
    Function: file.managed
        Name: /usr/local/bin/concourse_linux_amd64
      Result: False
     Comment: Specified sha256 checksum for /usr/local/bin/concourse_linux_amd64
 (87132e62f227e776b10178c21c83eb9f0049e4b53adc4e4dbb7eb06dfbb2e3b1) does not mat
ch actual checksum (e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852
b855)
     Started: 18:12:34.284629
    Duration: 865.019 ms
     Changes:
              ----------
              diff:
                  New file
```

Things to notice:
* `wget http://fileserver.maurusnet.test/downloads/concourse/concourse_linux_amd64` downloads the file from the local server **just fine**, so the local server works, is reachable and serves the file correctly over http. 
* Using the original github.com download URL with salt **works too**. 
* But downloading from the local Apache2 using `file.manage_file` fails by creating a 0 byte file instead, without any error reported.

### Steps to Reproduce Issue
1. Configure basic local apache2.
2. Try to download a file like this:

```bash
# this reports success, but really fails
$ salt-call file.manage_file /usr/local/bin/concourse_linux_amd64 '' '{}' 'http://fileserver.maurusnet.test/downloads/concourse/concourse_linux_amd64' "{'hash_type': 'sha256', 'hsum': '87132e62f227e776b10178c21c83eb9f0049e4b53adc4e4dbb7eb06dfbb2e3b1'}" 'root' 'root' '0755' base ''
local:
    ----------
    changes:
        ----------
        diff:
            New file
        group:
            root
        mode:
            0755
    comment:
        File /usr/local/bin/concourse_linux_amd64 updated
    name:
        /usr/local/bin/concourse_linux_amd64
    result:
        True
$ ls -al /usr/local/bin
total 92092
drwxrwsr-x  2 root    staff       4096 Nov  1 18:19 .
drwxrwsr-x 11 root    staff       4096 Sep 24 16:40 ..
-rwxr-xr-x  1 root    root           0 Nov  1 18:19 concourse_linux_amd64
# this reports success and works
$ salt-call file.manage_file /usr/local/bin/concourse_linux_amd64 '' '{}' 'https://github.com/concourse/concourse/releases/download/v3.6.0/concourse_linux_amd64' "{'hash_type': 'sha256', 'hsum': '87132e62f227e776b10178c21c83eb9f0049e4b53adc4e4dbb7eb06dfbb2e3b1'}" 'root' 'root' '0755' base ''
local:
    ----------
    changes:
        ----------
        diff:
            New file
        group:
            root
        mode:
            0755
    comment:
        File /usr/local/bin/concourse_linux_amd64 updated
    name:
        /usr/local/bin/concourse_linux_amd64
    result:
        True
```

### Versions Report
```
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: 1.9.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.2
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-3-amd64
         system: Linux
        version: debian 9.2
```

**Comment by Ch3LL** [comment]
can you give https://github.com/saltstack/salt/pull/44016 a try?

**Comment by dmurphy18** [comment]
This should be fixed by https://github.com/saltstack/salt/pull/43956 which will be in the next point release

**Comment by jdelic** [comment]
I can't run any non-release salt versions on these boxes at the moment, but I can validate that #44016 fixes this issue. Basically, setting:

```
<VirtualHost *:80>
    ServerName fileserver.maurusnet.test
    DocumentRoot /srv/aptly/public
    Alias /downloads /srv/downloads
    IndexOptions FancyIndexing
    <Directory /srv/downloads/>
        Require all granted
        Options Indexes
        ######
        ForceType application/octet-stream
        ###### ^^^ validate fix
    </Directory>
</VirtualHost>
```

immediately fixes the problem, so an empty `Content-type` HTTP header caused it.

Thanks for the immediate response. I'm closing this, as it's effectively addressed by a merged patch and thereby "duplicates" #44013.
